src/org/rascalmpl/uri/URIResolverRegistry.java

```diff
@@ -882,7 +883,27 @@ public FileChannel getReadableFileChannel(ISourceLocation uri) throws IOExceptio
 		return resolver.getReadableFileChannel(uri);
 	}
 
+	public Charset detectCharset(ISourceLocation sloc) {
```

Comment:
I'm not sure we should expose this. Normally you would call for a `getReader()` and just get it with the proper charset.

But it would avoid calling `getInputStream` twice.

Reply:
> this was actually abstracted away, as soon as you get a reader, we would either ask it from the scheme, and if they couldn't do it, use detect, without opening the stream twice. Is there no way to give the XML library a Reader instead of a InputStream?

I copied the method from one class to the other. There was no abstraction. This is about the writer, not the reader. Maybe that is a source of confusion?

Reply:
how can you detect the charset of the writer if the target file doesn't exist yet? I don't think it makes sense to detect it at write time. For writing it should be an argument that's passed in (most likely with a default of utf8). 

Even for reading I think it would be better to not have this detection by default anymore, but that would break existing code.

Reply:
`writeFile` used this private method to find out if `inferCharSet` was true, what the charset should be from the previous contents. Especially in case of appending this is important. Now with writing XML files there are similar options to the writing methods (inferCharset and charset), and so we need the exact same functionality that was previously private for Prelude.java, now in lang/xml/IO.java

Reply:
aah yes, I forgot about appending in the middle, do we support that? can you safely append an xml file?

Reply:
because our writers support offsets and lengths, we can always write into the middle of a file. Given the proper origins that we provide now, it becomes doable to safely replace a node inside an existing file.

Reply:
> how can you detect the charset of the writer if the target file doesn't exist yet? I don't think it makes sense to detect > it at write time. For writing it should be an argument that's passed in (most likely with a default of utf8).

We designed this together for `writeFile`; people can provide `charset`, but if they don't and they have `inferCharset` to true, we try to learn it from the existing file. This way we don't magically change the charset to utf8 of existing files just because the user of `writeFile` did not provide a charset explicitly. Also, it is absolutely necessary for append and write-into-middle as you said.

Reply:
true, I forgot about appending files. This would be the first place outside of `writeFile` where we actually support writes to offsets in the file. if you feel we need to add that for this library, then I understand this change. A pitty we cannot make that functionality more generic, although for a lot of writers, we can't really guarantee any accurate behavior.

We should have at least one test that does this writing in the middle, and verify it's still a correct xml. (and not some kind of nested xml header in there.

Reply:
and it's also about writing the file back the way you found it, in terms of encoding, without having to think about it.